### ISSUE TITLE
security/acme-client: may revert OPNsense config to old version

### ISSUE THREAD

**Issue by fraenki** [opnsense/plugins#1526]
When using the Let's Encrypt plugin it is possible to loose (unrelated) configuration changes. This may happen while any certificate operation is running, i.e. signing, renewing or revoking a certificate.

I think the issue is caused by the way the `certhelper.php` script is designed. When the script is started, it accesses the current OPNsense configuration to get all required information for the scheduled task. After performing the task, `certhelper.php` uses the same configuration object to write it's changes to the OPNsense configuration. This effectively reverts the configuration to the state when the script was started. All configuration changes that took place while `certhelper.php` was running get lost.

The problem can easily be reproduced when using Let's Encrypt with DNS-01 and **setting a long DNS sleep time** (i.e. 600 seconds). Then while acme.sh is sleeping (waiting for DNS updates), perform any configuration change in OPNsense, i.e. add a new firewall alias. Once acme.sh continues it's work and `certhelper.php` updates the OPNsense configuration, the new firewall alias is gone.

https://github.com/opnsense/plugins/blob/d313882e24d510cbc75411dddc071a7c7cade7a9/security/acme-client/src/opnsense/scripts/OPNsense/AcmeClient/certhelper.php#L121
https://github.com/opnsense/plugins/blob/d313882e24d510cbc75411dddc071a7c7cade7a9/security/acme-client/src/opnsense/scripts/OPNsense/AcmeClient/certhelper.php#L1262-L1264
https://github.com/opnsense/plugins/blob/d313882e24d510cbc75411dddc071a7c7cade7a9/security/acme-client/src/opnsense/scripts/OPNsense/AcmeClient/certhelper.php#L1422-L1425

@fichtner I'd assume this is a common problem and I'm doing something particular stupid in `certhelper.php` (besides writing spaghetti code, of course), but I'm not sure how this should be solved. Do you have some advice? Thanks :)

**Comment by blablup** [comment]
Maybe it would be possible to update the config via API?
Or to minimize the risk maybe it is possible to reload the config before doing an update. 
Depends what you plan to do, I could help a little (if needed and desired).

**Comment by fichtner** [comment]
Maybe @AdSchellevis can take a look. I'm a bit lost on the details.

**Comment by AdSchellevis** [comment]
It's not really a beauty indeed, but if you move the forceReload from here 
https://github.com/opnsense/plugins/blob/d313882e24d510cbc75411dddc071a7c7cade7a9/security/acme-client/src/opnsense/scripts/OPNsense/AcmeClient/certhelper.php#L1425

to somewhere above here 
https://github.com/opnsense/plugins/blob/d313882e24d510cbc75411dddc071a7c7cade7a9/security/acme-client/src/opnsense/scripts/OPNsense/AcmeClient/certhelper.php#L1414

It would likely improve the situation.... 

Locking the config while doing these updates is a possible other improvement (see lock() on the config object), I would also suggest to save outside the loop.

**Comment by karelkryda** [comment]
Hi @fraenki,
I would like to report that I have also encountered this problem with OPNsense version `23.7.9` and the `os-acme-client` plugin version `3.19`.
If I changed any settings (whether plugin, account or certificate settings) during a certificate request that was running for a long time, this change was always discarded and reverted to the original settings. I found this issue related to this behavior and would like to ask why this behavior still occurs. Should I open a new issue for this or continue here?

Thanks in advance for the information
